blog/content/post/2024-08-19-secureboot-freebsd.md

```diff
@@ -0,0 +1,378 @@
+Linux kernel, also ensuring it is signed and trusted.
+
+### UEFI Secure Boot status on FreeBSD - brief summary
+
+In short, the [FreeBSD wiki](https://wiki.freebsd.org/SecureBoot) proposes two
+stages of Secure Boot implementation. 
+
+First would be to enable booting the system with Microsoft-signed shim and a 
+FreeBSD-signed EFI loader, while also offering users to generate their own 
+keys and certificates.
+
+The second stage would involve having all of the drivers and kernel modules to
+be signed and authorized, resulting in a locked-down system akin to Fedora's 
+implementation. 
+
+As of the time of writing this post, the first stage is mostly complete. It is
+possible to create and sign a complete EFI executable containing the bootloader
+and the FreeBSD kernel, which can be booted using Dasharo firmware Secure Boot.
+
+Now, let's analyze what the said executable consists of.
+
+#### FreeBSD UEFI boot process
+
+The regular UEFI boot process on FreeBSD is as follows:
+
+* UEFI firmware loads `boot1.efi`
+* `boot1.efi` in turn loads `loader.efi`
+* Finally, `loader.efi` loads the kernel.
+
+The goal scenario, according to the FreeBSD Foundation, would be that a
+cryptographic handshake would verify each transition within the process. That
+is still under development, and the currently available option is a little less
+modular.
+
+It is possible to skip `boot1.efi` and combine the loader and the kernel into 
+one large package, which then can be signed with self-issued keys. The keys
+are enrolled into Dasharo firmware, which boots the loader-kernel object
+directly.
+
+## Creating a Secure Boot - ready FreeBSD EFI executable
+
+This guide is about migrating your current installation, not from scratch
+
+### Determine the loader of choice
+
+There are currently several flavors of the `loader.efi` available. The original
+one was written in 
+[FORTH](https://hackaday.com/2017/01/27/forth-the-hackers-language/).
+The 14.1 release defaults to a more modern Lua one, the final option being
+`loader_simp` - a simplified C implementation. This guide will only provide
+instructions on using the legacy FORTH loader and the current default.
+
+### Copy the boot filesystem
+
+```
+$ mkdir ~/bootfs
+$ mkdir ~/bootfs/boot
+$ cd ~/bootfs/boot
+$ cp -r /boot/kernel .
+$ cp -r /boot/defaults .
+$ cp /boot/loader.conf .
+$ cp /boot/*.rc .
+$ cp /boot/device.hints .
+$ cp /boot/loader.help* .
+```
+
+#### Lua loader
+
+```
+$ mkdir lua
+$ cp -r /boot/lua ./lua
+```
+
+#### FORTH loader
+
+```
+$ cp /boot/*.4th .
+```
+
+### Copy the fstab
+
+```
+$ mkdir ~/bootfs/etc
+$ cp /etc/fstab ~/bootfs/etc/fstab
+```
+
+### Create an image of the filesystem
+
+```
+$ cd ~/
+$ makefs bootfs.img bootfs
+```
+
+determine the size of the boot filesystem. This will determine how much memory 
+to reserve in the loader:
+
+```
+$ ls -l bootfs.img | awk '{print $5}'
+```
+add a safety factor of a couple hundred bytes (~512) to this number, and record
```

Comment:
Probably needs an empty line, also I would not break sentence by a block of code. Applies to other cases as well.

Reply:
[Fixed](https://github.com/3mdeb/news-and-ideas/pull/603/commits/1f14bdc84a7b4847b01bb79417f61d5d05422bf2#diff-1e142bc7519e97361fc0d3867abcb7fa3d257a4b5ccb79795e5e00acebf7d22fR152)